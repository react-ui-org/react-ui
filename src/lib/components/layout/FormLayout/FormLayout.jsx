import flattenChildren from 'react-keyed-flatten-children';
import PropTypes from 'prop-types';
import React from 'react';
import styles from './FormLayout.scss';

const PREDEFINED_LABEL_WIDTH_VALUES = ['auto', 'default', 'limited'];

const FormLayout = (props) => {
  const {
    children,
    fieldLayout,
    id,
    labelWidth,
    labelWidthFallback,
  } = props;

  if (!children) {
    return null;
  }

  const HAS_CUSTOM_LABEL_WIDTH = !PREDEFINED_LABEL_WIDTH_VALUES.includes(labelWidth);
  let customLabelWidth;

  if (HAS_CUSTOM_LABEL_WIDTH) {
    customLabelWidth = labelWidth;
  }

  const fieldLayoutClass = (layout) => {
    if (layout === 'horizontal') {
      return styles.rootFieldLayoutHorizontal;
    }

    return styles.rootFieldLayoutVertical;
  };

  const labelWidthClass = (width) => {
    if (HAS_CUSTOM_LABEL_WIDTH) {
      return styles.hasRootLabelWidthCustom;
    }

    if (width === 'auto') {
      return styles.hasRootLabelWidthAuto;
    }

    if (width === 'limited') {
      return styles.hasRootLabelWidthLimited;
    }

    return styles.hasRootLabelWidthDefault;
  };

  const inlineStyle = (customWidth, customWidthFallback) => {
    const style = {};

    if (customWidth) {
      style['--rui-custom-label-width'] = customWidth;
    }

    if (customWidthFallback) {
      style['--rui-custom-label-width-fallback'] = customWidthFallback;
    }

    return style;
  };

  return (
    <div
      id={id}
      className={[
        styles.root,
        fieldLayoutClass(fieldLayout),
        fieldLayout === 'horizontal' ? labelWidthClass(labelWidth) : '',
        labelWidthFallback ? styles.hasRootCustomLabelWidthFallback : '',
      ].join(' ')}
      style={inlineStyle(customLabelWidth, labelWidthFallback)}
    >
      {flattenChildren(children).map((child) => {
        if (!React.isValidElement(child)) {
          return null;
        }

        return React.cloneElement(child, {
          inFormLayout: true,
          layout: fieldLayout,
        });
      })}
    </div>
  );
};

FormLayout.defaultProps = {
  children: null,
  fieldLayout: 'vertical',
  id: undefined,
  labelWidth: 'default',
  labelWidthFallback: undefined,
};

FormLayout.propTypes = {
  children: PropTypes.node,
  fieldLayout: PropTypes.oneOf(['horizontal', 'vertical']),
  id: PropTypes.string,
  labelWidth: PropTypes.oneOfType([
    PropTypes.oneOf(PREDEFINED_LABEL_WIDTH_VALUES),
    PropTypes.string,
  ]),
  labelWidthFallback: PropTypes.string,
};

export default FormLayout;
