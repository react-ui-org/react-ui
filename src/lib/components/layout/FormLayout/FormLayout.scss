// 1. Form field label width is defined using custom property. Form Layout variants are created just
//    by adjusting this value, either through a particular modification class, or manually through
//    a prop. Form fields then inherit the layout using the `subgrid` feature which makes it possible
//    to turn on automatic size of grid column with labels (in supporting browsers). There is always
//    a fallback value for browsers that don't support `subgrid` (see `tools/forms/_layouts.scss`).
//    This value is passed down to form fields utilising custom property cascade.
// 2. By default, form field label width is set to a configurable global value.
// 3. Custom label width can be entered individually through a prop for each use of Form Layout.
// 4. Form Layout can also automatically adjust the width of column with labels according to the
//    longest label (which is possible only with `subgrid`). By default, automatic width is limited
//    to 50 % of available horizontal space. However, the automatic width is currently supported
//    only in Firefox so a fallback value is necessary (a global default which can be customised per
//    use).

@import '../../../styles/settings/forms';
@import '../../../styles/settings/forms-theme';
@import '../../../styles/settings/layouts';
@import '../../../styles/tools/breakpoints';
@import '../../../styles/tools/offset';

.root {
  margin-bottom: $layout-common-bottom-offset;
}

.rootFieldLayoutVertical,
.rootFieldLayoutHorizontal {
  display: grid;
  grid-template-columns: 1fr;
  grid-row-gap: $form-field-vertical-outer-spacing;
}

.rootFieldLayoutHorizontal {
  @include breakpoint-up($form-field-horizontal-breakpoint) {
    grid-template-columns: var(--rui-local-label-width) 1fr; // 1.
  }
}

.hasRootLabelWidthDefault {
  --rui-local-label-width: #{$form-layout-horizontal-label-default-width}; // 1., 2.
  --rui-local-label-width-fallback: #{$form-layout-horizontal-label-default-width}; // 1., 2.
}

.hasRootLabelWidthCustom {
  --rui-local-label-width: var(--rui-custom-label-width); // 3.
  --rui-local-label-width-fallback: var(--rui-custom-label-width); // 3.
}

.hasRootLabelWidthAuto {
  --rui-local-label-width: #{$form-layout-horizontal-label-auto-width}; // 4.
  --rui-local-label-width-fallback: #{$form-layout-horizontal-label-auto-width-fallback}; // 4.
}

.hasRootLabelWidthAuto.hasRootCustomLabelAutoWidthFallback {
  --rui-local-label-width-fallback: var(--rui-custom-label-auto-width-fallback); // 4.
}
