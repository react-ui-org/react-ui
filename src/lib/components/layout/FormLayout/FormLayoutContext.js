import React from 'react';

const FormLayoutContext = React.createContext({
  layout: undefined,
});

export default FormLayoutContext;
