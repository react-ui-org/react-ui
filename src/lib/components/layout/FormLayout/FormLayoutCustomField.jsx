import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../../provider';
import styles from './FormLayoutCustomField.scss';

export const FormLayoutCustomField = ({
  children,
  fullWidth,
  id,
  label,
  layout,
}) => (
  <div
    id={id}
    className={`
      ${styles.root}
      ${fullWidth ? styles.isRootFullWidth : ''}
      ${layout === 'vertical' ? styles.rootLayoutVertical : styles.rootLayoutHorizontal}
    `.trim()}
  >
    {label && (
      <div
        id={id && `${id}__label`}
        className={styles.label}
      >
        {label}
      </div>
    )}
    <div
      id={id && `${id}__field`}
      className={styles.field}
    >
      {children}
    </div>
  </div>
);

FormLayoutCustomField.defaultProps = {
  children: null,
  fullWidth: false,
  id: undefined,
  label: null,
  layout: 'vertical',
};

FormLayoutCustomField.propTypes = {
  /**
   * Custom HTML or React component(s).
   */
  children: PropTypes.node,
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Optional label of the field.
   */
  label: PropTypes.string,
  /**
   * Layout of the field, controlled by parent FormLayout.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
};

export const FormLayoutCustomFieldWithContext = withProviderContext(FormLayoutCustomField, 'FormLayoutCustomField');

export default FormLayoutCustomFieldWithContext;
