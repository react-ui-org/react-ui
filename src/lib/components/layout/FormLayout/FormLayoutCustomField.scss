@import '../../../styles/tools/forms/layouts';

.root {
  @include form-field-in-form-layout();
}

.label {
  overflow-wrap: break-word;
}

.rootLayoutHorizontal {
  @include form-field-layout-horizontal();
}

.rootLayoutVertical {
  @include form-field-layout-vertical();
}

.rootLayoutHorizontal .field,
.rootLayoutVertical .field {
  width: auto;
}

.isRootFullWidth .field {
  justify-self: stretch;
}

.root.rootLayoutHorizontal {
  @include breakpoint-up($form-field-horizontal-breakpoint) {
    grid-template-columns: var(--rui-local-label-width-fallback) 1fr;
    grid-template-columns: subgrid;
  }
}
