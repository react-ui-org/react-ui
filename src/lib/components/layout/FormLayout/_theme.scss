$row-gap: var(--rui-FormLayout__row-gap);
$horizontal-label-width: var(--rui-FormLayout--horizontal__label__width);
$horizontal-label-width-auto: var(--rui-FormLayout--horizontal__label__width--auto);
$horizontal-label-width-limited: var(--rui-FormLayout--horizontal__label__width--limited);
