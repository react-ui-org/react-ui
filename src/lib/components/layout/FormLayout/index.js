export { default as FormLayout } from './FormLayout';
export { FormLayoutContext } from './FormLayoutContext';
export { default as FormLayoutCustomField } from './FormLayoutCustomField';
