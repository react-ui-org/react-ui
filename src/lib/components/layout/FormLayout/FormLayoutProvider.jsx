import PropTypes from 'prop-types';
import React from 'react';
import FormLayoutContext from './FormLayoutContext';

const FormLayoutProvider = ({
  children,
  layout,
}) => (
  <FormLayoutContext.Provider
    value={{ layout }}
  >
    {children}
  </FormLayoutContext.Provider>
);

FormLayoutProvider.defaultProps = {
  children: null,
  layout: 'vertical',
};

FormLayoutProvider.propTypes = {
  children: PropTypes.node,
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
};

export default FormLayoutProvider;
