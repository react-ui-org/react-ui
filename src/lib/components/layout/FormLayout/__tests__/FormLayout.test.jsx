import React from 'react';
import { shallow } from 'enzyme';
import FormLayout from '../FormLayout';

describe('rendering', () => {
  it('renders correctly with a single child', () => {
    const tree = shallow((
      <FormLayout>
        <span>content</span>
      </FormLayout>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with multiple children', () => {
    const tree = shallow((
      <FormLayout>
        <span>content 1</span>
        <span>content 2</span>
        <span>content 3</span>
      </FormLayout>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with custom label width', () => {
    const tree = shallow((
      <FormLayout
        fieldLayout="horizontal"
        id="test-id"
        labelWidth="300px"
      >
        <span>content</span>
      </FormLayout>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow((
      <FormLayout
        fieldLayout="horizontal"
        id="test-id"
        labelAutoWidthFallback="200px"
        labelWidth="auto"
      >
        <span>content</span>
      </FormLayout>
    ));

    expect(tree).toMatchSnapshot();
  });
});
