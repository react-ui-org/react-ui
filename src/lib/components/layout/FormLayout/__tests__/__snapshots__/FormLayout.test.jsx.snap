// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with a single child 1`] = `
<div
  className="root rootFieldLayoutVertical  "
  style={Object {}}
>
  <span
    inFormLayout={true}
    key=".0"
    layout="vertical"
  >
    content
  </span>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root rootFieldLayoutHorizontal hasRootLabelWidthAuto hasRootCustomLabelAutoWidthFallback"
  id="test-id"
  style={
    Object {
      "--rui-custom-label-auto-width-fallback": "200px",
    }
  }
>
  <span
    inFormLayout={true}
    key=".0"
    layout="horizontal"
  >
    content
  </span>
</div>
`;

exports[`rendering renders correctly with custom label width 1`] = `
<div
  className="root rootFieldLayoutHorizontal hasRootLabelWidthCustom "
  id="test-id"
  style={
    Object {
      "--rui-custom-label-width": "300px",
    }
  }
>
  <span
    inFormLayout={true}
    key=".0"
    layout="horizontal"
  >
    content
  </span>
</div>
`;

exports[`rendering renders correctly with multiple children 1`] = `
<div
  className="root rootFieldLayoutVertical  "
  style={Object {}}
>
  <span
    inFormLayout={true}
    key=".0"
    layout="vertical"
  >
    content 1
  </span>
  <span
    inFormLayout={true}
    key=".1"
    layout="vertical"
  >
    content 2
  </span>
  <span
    inFormLayout={true}
    key=".2"
    layout="vertical"
  >
    content 3
  </span>
</div>
`;
